File: axiom/neuron/crf_loss.php
<?php
require_once '..\..\utility.php';
$i = 0;
$txt[$i++] = "\\({s}_{t} = \\sum\\limits_{i=1}^{t} {G}_{{y}_{i},{y}_{i - 1}} + \\sum\\limits_{i=0}^{t} {x}_{i,{y}_{i}}\\tag*{Eq.s_definition}\\)";
$txt[$i++] = "\\({z}_{t} = [{y}_{t}]\\sum_{{y}_{0:t}} {\\color{blue} e}^{- {s}_{t}}\\tag*{Eq.z_definition}\\)";
$txt[$i++] = "\\({x'}_{t} = - \\log{{z}_{t}}\\tag*{Eq.x_quote_definition}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} - \\log{\\sum\\limits_{\\substack{}} {\\color{blue} e}^{- {x'}_{t} - G}}\\tag*{Eq.plausible0}\\)";
$txt[$i++] = "\\(- \\log{\\left(\\frac{{\\color{blue} e}^{- {s}_{t}}}{\\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {s}_{t}}} \\right)} = {s}_{t} + \\log{\\sum\\limits_{\\substack{}} {\\color{blue} e}^{- {x'}_{t}}}\\tag*{Eq.plausible1}\\)";
$txt[$i++] = "\\({s}_{t + 1} - {s}_{t} = - \\sum\\limits_{i=1}^{t} {G}_{{y}_{i},{y}_{i - 1}} + \\sum\\limits_{i=1}^{t + 1} {G}_{{y}_{i},{y}_{i - 1}} - \\sum\\limits_{i=0}^{t} {x}_{i,{y}_{i}} + \\sum\\limits_{i=0}^{t + 1} {x}_{i,{y}_{i}}\\tag*{Eq[0]}\\)";
$txt[$i++] = "\\({s}_{t + 1} = {G}_{{y}_{t + 1},{y}_{t}} + {s}_{t} + {x}_{t + 1,{y}_{t + 1}}\\tag*{Eq[1]}\\)";
$txt[$i++] = "\\({z}_{t + 1} = [{y}_{t + 1}]\\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {s}_{t + 1}}\\tag*{Eq[2]}\\)";
$txt[$i++] = "\\({z}_{t + 1} = [{y}_{t + 1}]\\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {G}_{{y}_{t + 1},{y}_{t}} - {s}_{t} - {x}_{t + 1,{y}_{t + 1}}}\\tag*{Eq[3]}\\)";
$txt[$i++] = "\\({z}_{t + 1} = [{y}_{t + 1}]\\left({\\color{blue} e}^{- {x}_{t + 1,{y}_{t + 1}}} \\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {G}_{{y}_{t + 1},{y}_{t}}} {\\color{blue} e}^{- {s}_{t}}\\right)\\tag*{Eq[4]}\\)";
$txt[$i++] = "\\({z}_{t + 1} = {\\color{blue} e}^{- {x}_{t + 1}} [{y}_{t + 1}]\\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {G}_{{y}_{t + 1},{y}_{t}}} {\\color{blue} e}^{- {s}_{t}}\\tag*{Eq[5]}\\)";
$txt[$i++] = "\\({z}_{t + 1} = {\\color{blue} e}^{- {x}_{t + 1}} [{y}_{t + 1}]\\sum_{{y}_{t}} {\\color{blue} e}^{- {G}_{{y}_{t + 1},{y}_{t}}} \\sum_{{y}_{0:t}} {\\color{blue} e}^{- {s}_{t}}\\tag*{Eq[6]}\\)";
$txt[$i++] = "\\({z}_{t + 1} = {\\color{blue} e}^{- {x}_{t + 1}} [{y}_{t + 1}]\\sum\\limits_{\\substack{}} [{y}_{t}]\\left({\\color{blue} e}^{- {G}_{{y}_{t + 1},{y}_{t}}} \\sum_{{y}_{0:t}} {\\color{blue} e}^{- {s}_{t}}\\right)\\tag*{Eq[7]}\\)";
$txt[$i++] = "\\({z}_{t + 1} = \\left([{y}_{t}]\\sum_{{y}_{0:t}} {\\color{blue} e}^{- {s}_{t}} \\times {\\color{blue} e}^{- G^{\\color{red} T}}\\right) {\\color{blue} e}^{- {x}_{t + 1}}\\tag*{Eq[8]}\\)";
$txt[$i++] = "\\({z}_{t + 1} = \\left({z}_{t} \\times {\\color{blue} e}^{- G^{\\color{red} T}}\\right) {\\color{blue} e}^{- {x}_{t + 1}}\\tag*{Eq[9]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = - \\log{{z}_{t + 1}}\\tag*{Eq[10]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = - \\log{\\left(\\left({z}_{t} \\times {\\color{blue} e}^{- G^{\\color{red} T}}\\right) {\\color{blue} e}^{- {x}_{t + 1}} \\right)}\\tag*{Eq[11]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} - \\log{\\left({z}_{t} \\times {\\color{blue} e}^{- G^{\\color{red} T}} \\right)}\\tag*{Eq[12]}\\)";
$txt[$i++] = "\\({z}_{t} = {\\color{blue} e}^{- {x'}_{t}}\\tag*{Eq.z_definition_by_x_quote}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} - \\log{\\left({\\color{blue} e}^{- {x'}_{t}} \\times {\\color{blue} e}^{- G^{\\color{red} T}} \\right)}\\tag*{Eq[13]}\\)";
$txt[$i++] = "\\({x'}_{t + 1} = {x}_{t + 1} - \\log{\\sum\\limits_{\\substack{}} {\\color{blue} e}^{- {x'}_{t} - G}}\\tag*{Eq.plausible0}\\)";
$txt[$i++] = "\\(\\log{\\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {s}_{t}}} = \\log{\\sum\\limits_{\\substack{}} {\\color{blue} e}^{- {x'}_{t}}}\\tag*{Eq[14]}\\)";
$txt[$i++] = "\\(\\sum\\limits_{\\substack{}} {\\color{blue} e}^{- {x'}_{t}} = \\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {s}_{t}}\\tag*{Eq[15]}\\)";
$txt[$i++] = "\\(\\sum\\limits_{\\substack{}} {z}_{t} = \\sum\\limits_{\\substack{}} [{y}_{t}]\\sum_{{y}_{0:t}} {\\color{blue} e}^{- {s}_{t}}\\tag*{Eq[16]}\\)";
$txt[$i++] = "\\(\\sum\\limits_{\\substack{}} {z}_{t} = \\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {s}_{t}}\\tag*{Eq[17]}\\)";
$txt[$i++] = "\\(\\sum\\limits_{\\substack{}} {\\color{blue} e}^{- {x'}_{t}} = \\sum_{{y}_{0:t + 1}} {\\color{blue} e}^{- {s}_{t}}\\tag*{Eq[15]}\\)";
render(__FILE__, $txt);
?>        

File: axiom/neuron/crf_loss.py
from sympy.core.symbol import Symbol
from sympy.utility import Ref, Sum, plausible
from sympy.core.relational import Equality
from sympy.tensor.indexed import IndexedBase
import sympy
from sympy import log, exp
from sympy.core.numbers import oo


@plausible
def apply(G, x, y):
    _, d = x.shape

    i = Symbol('i', integer=True)
    t = Symbol('t', integer=True, nonnegative=True)

    s = IndexedBase('s', (oo,),
                    definition=Ref[t](Sum[i:1:t](G[y[i], y[i - 1]]) + Sum[i:0:t](x[i, y[i]])))

    z = IndexedBase('z', shape=(oo, d),
                    definition=Ref[t](Ref[y[t]](Sum[y[0:t]](sympy.E ** -s[t]))))

    x_quote = IndexedBase("x'", shape=(oo, d),
                    definition=-Ref[t](sympy.log(z[t])))

    return Equality(x_quote[t + 1], -log(Sum(exp(-x_quote[t] - G))) + x[t + 1]), \
        Equality(-log(exp(-s[t]) / Sum[y[:t + 1]](exp(-s[t]))), log(Sum(exp(-x_quote[t]))) + s[t])


from sympy.utility import check


@check
def prove(Eq):
    d = Symbol('d', integer=True)
    G = IndexedBase('G', (d, d))
    x = IndexedBase('x', (oo, d))
    y = IndexedBase('y', (oo,), integer=True)

    # n is the length of the sequence
    # d is the number of output labels

    Eq.s_definition, Eq.z_definition, Eq.x_quote_definition, Eq.plausible0, Eq.plausible1 = apply(G, x, y)
    
    t = Eq.plausible0.lhs.indices[0] - 1
    x_quote = Eq.plausible0.lhs.base

    z = x_quote.definition.args[1].function.arg.base
    s = z.definition.function.function.function.arg.args[1].base
    
    Eq << Eq.s_definition.subs(t, t + 1) - Eq.s_definition

    Eq << Eq[-1].this.rhs.simplify() + s[t]

    Eq << Eq.z_definition.subs(t, t + 1)
    
    Eq << Eq[-1].this.rhs.subs(Eq[-2])

    Eq << Eq[-1].this.rhs.function.simplify()

    Eq << Eq[-1].this.rhs.as_two_terms()

    Eq << Eq[-1].this.rhs.args[1].function.bisect(back=1)

    Eq << Eq[-1].this.rhs.args[1].function.as_Ref()

    Eq << Eq[-1].this.rhs.args[1].function.function.as_two_terms()

    Eq << Eq[-1].this.rhs.subs(Eq.z_definition.reversed)

    Eq << Eq.x_quote_definition.subs(t, t + 1)

    Eq << Eq[-1].this.rhs.subs(Eq[-2])

    Eq << Eq[-1].this.rhs.args[1].as_Add()

    Eq.z_definition_by_x_quote = sympy.E ** -Eq.x_quote_definition.reversed

    Eq << Eq[-1].subs(Eq.z_definition_by_x_quote)

    Eq << Eq[-1].this.rhs.args[1].args[1].arg.simplify()

    Eq << Eq.plausible1.this.lhs.args[1].as_Add()

    Eq << Eq[-1].exp().reversed

    Eq << Eq.z_definition.summation()

    Eq << Eq[-1].this.rhs.as_Sum()

    Eq << Eq[-1].subs(Eq.z_definition_by_x_quote)


# reference: Neural Architectures for Named Entity Recognition.pdf
if __name__ == '__main__':
    prove(__file__)
